File: leetcode/count-vowels-permutation.py
# 1220. Count Vowels Permutation
# 🔴 Hard
#
# https://leetcode.com/problems/count-vowels-permutation/
#
# Tags: Dynamic Programming

import timeit

MOD = 10**9 + 7

# Let dp[i][j] be the number of strings of length i that ends with the
# j-th vowel, and we know that:
#
# Each character is a lower case vowel ('a', 'e', 'i', 'o', 'u')
# Each vowel 'a' may only be followed by an 'e'.
# Each vowel 'e' may only be followed by an 'a' or an 'i'.
# Each vowel 'i' may not be followed by another 'i'.
# Each vowel 'o' may only be followed by an 'i' or a 'u'.
# Each vowel 'u' may only be followed by an 'a'.
#
# That means that we can iterate over all the numbers smaller than n
# and, for each iteration, the number of vowel combinations that we
# can do for each vowel will be the sum of the possible combinations
# of the vowels that are allowed to go before it according to the rules.
#
# Time complexity: O(n) - We loop once for each value between 1..n
# Space complexity: O(1) - We use constant space.
#
# Runtime: 326 ms, faster than 75.50%
# Memory Usage: 14.2 MB, less than 68.69%
class DP:
    def countVowelPermutation(self, n: int) -> int:
        # Create the dp array and initialize it with the number of
        # possible ways to form 1 character long strings that finish in
        # the given vowel, 0 => a, 1 => e, 2 => i, 3 => o, 4 => u
        prev = [1] * 5
        # Initialize an array where we will put the computed values on
        # each iteration.
        curr = prev.copy()
        for _ in range(n - 1):
            # a can follow e, i, u
            curr[0] = prev[1] + prev[2] + prev[4]
            # e can follow a, i
            curr[1] = prev[0] + prev[2]
            # i can follow e, o
            curr[2] = prev[1] + prev[3]
            # o can follow i
            curr[3] = prev[2]
            # u can follow i, o
            curr[4] = prev[2] + prev[3]
            # Substitute the previous array for the new one.
            prev = curr.copy()

        # The result is the sum of the different ways to combine the
        # vowels.
        return sum(curr) % MOD


# Same solution as above but update all the values on one assignment.
#
# Time complexity: O(n) - We loop once for each value between 1..n
# Space complexity: O(1) - We use constant space.
#
# Runtime: 311 ms, faster than 77.53%
# Memory Usage: 14.1 MB, less than 80.85%
class DPSingleList:
    def countVowelPermutation(self, n: int) -> int:
        a = e = i = o = u = 1
        for _ in range(n - 1):
            a, e, i, o, u = e + i + u, a + i, e + o, i, i + o
        return (a + e + i + o + u) % MOD


def test():
    executors = [
        DP,
        DPSingleList,
    ]
    tests = [
        [1, 5],
        [2, 10],
        [5, 68],
        [2345, 451992221],
    ]
    for executor in executors:
        start = timeit.default_timer()
        for _ in range(int(float("1"))):
            for n, t in enumerate(tests):
                sol = executor()
                result = sol.countVowelPermutation(t[0])
                exp = t[1]
                assert result == exp, (
                    f"\033[93m» {result} <> {exp}\033[91m for "
                    + f"test {n} using \033[1m{executor.__name__}"
                )
        stop = timeit.default_timer()
        used = str(round(stop - start, 5))
        cols = "{0:20}{1:10}{2:10}"
        res = cols.format(executor.__name__, used, "seconds")
        print(f"\033[92m» {res}\033[0m")


test()

File: leetcode/count-vowels-permutation.rs
// 1220. Count Vowels Permutation
// 🔴 Hard
//
// https://leetcode.com/problems/count-vowels-permutation/
//
// Tags: Dynamic Programming

struct Solution;
impl Solution {
    /// For each length of the string we are creating, we can compute the number
    /// of ways that we can build it using the number of ways in which we built
    /// the string that is one character shorter for each of the 5 possible ways
    /// in which it can finish, then using that value and adding the vowels that
    /// we are allowed to for each ending.
    ///
    /// Time complexity: O(n) - One loop that runs n-1 times and does constant
    /// time work inside the loop.
    /// Space complexity: O(1) - We use one array of 5 elements of extra memory
    /// that will be stored in the stack.
    ///
    /// Runtime 1 ms Beats 100%
    /// Memory 1.99 MB Beats 90.91%
    pub fn count_vowel_permutation(n: i32) -> i32 {
        let modu = 1_000_000_007;
        let mut dp = [1i64; 5];
        for _ in 0..n - 1 {
            (dp[0], dp[1], dp[2], dp[3], dp[4]) = (
                (dp[1] + dp[2] + dp[4]) % modu,
                (dp[0] + dp[2]) % modu,
                (dp[1] + dp[3]) % modu,
                dp[2],
                (dp[2] + dp[3]) % modu,
            );
        }
        (dp.iter().sum::<i64>() % modu) as i32
    }
}

// Tests.
fn main() {
    let tests = [(1, 5), (2, 10), (5, 68), (20000, 759959057)];
    for t in tests {
        assert_eq!(Solution::count_vowel_permutation(t.0), t.1);
    }
    println!("\x1b[92m» All tests passed!\x1b[0m")
}
